src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexerOutpointsRepository.cs

```diff
@@ -95,6 +101,44 @@ public void SaveAllItems()
             }
         }
 
+        /// <summary>Persists rewind data into the repository.</summary>
+        /// <param name="rewindData">The data to be persisted.</param>
+        public void RecordRewindData(AddressIndexerRewindData rewindData)
+        {
+            lock (this.LockObject)
+            {
+                this.addressIndexerRewindData.Insert(rewindData);
+            }
+        }
+
+        /// <summary>Deletes rewind data that is more than maxReorg blocks old.</summary>
+        /// <param name="height">The threshold below which data will be deleted.</param>
+        public void PurgeOldRewindData(int height)
+        {
+            lock (this.LockObject)
+            {
+                // Generally there will only be one result here at most, as this should be getting called once per block.
+                foreach (AddressIndexerRewindData rewindData in this.addressIndexerRewindData.Find(Query.LT("BlockHeightIndex", height)))
+                {
+                    this.addressIndexerRewindData.Delete(rewindData.BlockHash);
+                }
+            }
+        }
+
+        public void Rewind(uint256 blockHash)
+        {
+            lock (this.LockObject)
+            {
+                AddressIndexerRewindData rewindData = this.addressIndexerRewindData.FindById(blockHash.ToString());
+
+                // Put the spent outputs back into the cache.
+                foreach (OutPointData outPointData in rewindData.SpentOutputs)
+                    this.AddOutPointData(outPointData);
```

Comment:
Deadlock. 

You call `AddOutPointData` while holding a lock, but inside of the AddOutPointData we call AddOrUpdate which then acquires the lock. 

Basically that means that PR is not tested because rewind doesnt work.